File: stacks/nodejs/react-fastify/rest/bff/src/flow/github/adapter.test.ts
import assert from "node:assert/strict";
import fs from "node:fs";
import path from "node:path";
import { test } from "node:test";
import YAML from "yaml";
import type { ProviderAdapterInput } from "../types";
import { buildGitHubProviderInput } from "./adapter";
import { blockedOnReviewSource, trunkIntegrationFailureSource } from "./fixtures";

function findRepoRoot(startDir: string): string {
  let current = startDir;
  while (true) {
    if (fs.existsSync(path.join(current, "schema"))) {
      return current;
    }
    const parent = path.dirname(current);
    if (parent === current) {
      throw new Error("Repository root not found from adapter tests");
    }
    current = parent;
  }
}

const repoRoot = findRepoRoot(__dirname);
const fixtureDir = path.join(repoRoot, "schema", "fixtures", "provider-adapter-input");

function loadNormalizedFixture(filename: string): ProviderAdapterInput {
  const filePath = path.join(fixtureDir, filename);
  const content = fs.readFileSync(filePath, "utf8");
  const parsed = YAML.parse(content) as Partial<ProviderAdapterInput> & Record<string, unknown>;
  return {
    repository: parsed.repository ?? { provider: "github", provider_id: "", full_name: "", default_branch: "", fetched_at: "" },
    changes: parsed.changes ?? [],
    actors: parsed.actors ?? [],
    review_states: parsed.review_states ?? [],
    validation_runs: parsed.validation_runs ?? [],
    merge_events: parsed.merge_events ?? [],
    ownership_hints: parsed.ownership_hints ?? [],
  };
}

function pruneUndefined<T>(value: T): T {
  if (Array.isArray(value)) {
    return value.map((item) => pruneUndefined(item)) as T;
  }
  if (value && typeof value === "object") {
    const entries = Object.entries(value as Record<string, unknown>).filter(
      ([, v]) => v !== undefined,
    );
    const result: Record<string, unknown> = {};
    for (const [key, val] of entries) {
      result[key] = pruneUndefined(val);
    }
    return result as T;
  }
  return value;
}

function normalizeOutput(input: ProviderAdapterInput): ProviderAdapterInput {
  const normalized = pruneUndefined<ProviderAdapterInput>({
    ...input,
    changes: [...input.changes].sort((a, b) => a.provider_id.localeCompare(b.provider_id)),
    actors: [...input.actors].sort((a, b) => a.provider_id.localeCompare(b.provider_id)),
    review_states: [...input.review_states].sort((a, b) => a.change_id.localeCompare(b.change_id)),
    validation_runs: [...input.validation_runs].sort((a, b) => {
      if (a.change_id !== b.change_id) {
        return a.change_id.localeCompare(b.change_id);
      }
      if (a.scope !== b.scope) {
        return a.scope.localeCompare(b.scope);
      }
      return (a.run_at ?? "").localeCompare(b.run_at ?? "");
    }),
    merge_events: [...input.merge_events].sort((a, b) => a.change_id.localeCompare(b.change_id)),
    ownership_hints: [...input.ownership_hints].sort((a, b) => {
        const pathA = a.path_pattern ?? "";
        const pathB = b.path_pattern ?? "";
        if (pathA !== pathB) {
          return pathA.localeCompare(pathB);
        }
        return (a.source ?? "").localeCompare(b.source ?? "");
      }),
  });

  return normalized;
}

test("normalizes GitHub data for blocked-on-review scenario", () => {
  const result = buildGitHubProviderInput(blockedOnReviewSource);
  const expected = loadNormalizedFixture("blocked-on-review-github.yaml");

  assert.deepStrictEqual(normalizeOutput(result), normalizeOutput(expected));
});

test("normalizes GitHub data with trunk integration failure after merge", () => {
  const result = buildGitHubProviderInput(trunkIntegrationFailureSource);
  const expected = loadNormalizedFixture("trunk-integration-failed-github.yaml");

  assert.deepStrictEqual(normalizeOutput(result), normalizeOutput(expected));
});

File: stacks/go/net-http/rest/bff/flow/github/adapter_test.go
package github

import (
	"os"
	"path/filepath"
	"reflect"
	"runtime"
	"sort"
	"testing"

	"gopkg.in/yaml.v3"

	"idp-go-net-http-rest/bff/flow"
)

func loadFixture(t *testing.T, name string) flow.ProviderAdapterInput {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatalf("unable to resolve caller")
	}
	base := filepath.Dir(file)
	// stacks/go/net-http/rest/bff/flow/github -> repo root is six levels up.
	path := filepath.Join(base, "..", "..", "..", "..", "..", "..", "..", "schema", "fixtures", "provider-adapter-input", name+".yaml")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	var input flow.ProviderAdapterInput
	if err := yaml.Unmarshal(data, &input); err != nil {
		t.Fatalf("unmarshal fixture %s: %v", name, err)
	}
	if input.EvidenceStates == nil {
		input.EvidenceStates = []flow.NormalizedEvidenceState{}
	}
	if input.MergeEvents == nil {
		input.MergeEvents = []flow.NormalizedMergeEvent{}
	}
	return input
}

func sortInput(input flow.ProviderAdapterInput) flow.ProviderAdapterInput {
	out := input
	out.Changes = append([]flow.NormalizedChange(nil), input.Changes...)
	sort.SliceStable(out.Changes, func(i, j int) bool { return out.Changes[i].ProviderID < out.Changes[j].ProviderID })

	out.Actors = append([]flow.NormalizedActor(nil), input.Actors...)
	sort.SliceStable(out.Actors, func(i, j int) bool { return out.Actors[i].ProviderID < out.Actors[j].ProviderID })

	out.ReviewStates = append([]flow.NormalizedReviewState(nil), input.ReviewStates...)
	sort.SliceStable(out.ReviewStates, func(i, j int) bool { return out.ReviewStates[i].ChangeID < out.ReviewStates[j].ChangeID })

	out.ValidationRuns = append([]flow.NormalizedValidationRun(nil), input.ValidationRuns...)
	sort.SliceStable(out.ValidationRuns, func(i, j int) bool {
		a := out.ValidationRuns[i]
		b := out.ValidationRuns[j]
		if a.ChangeID != b.ChangeID {
			return a.ChangeID < b.ChangeID
		}
		if a.Scope != b.Scope {
			return a.Scope < b.Scope
		}
		return a.RunAt < b.RunAt
	})

	out.MergeEvents = append([]flow.NormalizedMergeEvent(nil), input.MergeEvents...)
	sort.SliceStable(out.MergeEvents, func(i, j int) bool { return out.MergeEvents[i].ChangeID < out.MergeEvents[j].ChangeID })

	out.OwnershipHints = append([]flow.NormalizedOwnershipHint(nil), input.OwnershipHints...)
	sort.SliceStable(out.OwnershipHints, func(i, j int) bool {
		a := out.OwnershipHints[i]
		b := out.OwnershipHints[j]
		if a.PathPattern != b.PathPattern {
			return a.PathPattern < b.PathPattern
		}
		return a.Source < b.Source
	})

	for i := range out.Changes {
		if out.Changes[i].WorkItemRefs == nil {
			out.Changes[i].WorkItemRefs = []flow.NormalizedWorkItemRef{}
		}
	}

	for i := range out.OwnershipHints {
		if out.OwnershipHints[i].OwnerActorIDs == nil {
			out.OwnershipHints[i].OwnerActorIDs = []string{}
		}
		if out.OwnershipHints[i].OwnerTeamNames == nil {
			out.OwnershipHints[i].OwnerTeamNames = []string{}
		}
	}

	if out.MergeEvents == nil {
		out.MergeEvents = []flow.NormalizedMergeEvent{}
	}
	if out.EvidenceStates == nil {
		out.EvidenceStates = []flow.NormalizedEvidenceState{}
	}
	return out
}

func assertProviderInputEqual(t *testing.T, got, want flow.ProviderAdapterInput) {
	t.Helper()
	gotS := sortInput(got)
	wantS := sortInput(want)
	if !reflect.DeepEqual(gotS, wantS) {
		t.Fatalf("provider input mismatch\n got: %#v\nwant: %#v", gotS, wantS)
	}
}

func TestGitHubAdapterBlockedOnReview(t *testing.T) {
	got := BuildProviderInput(blockedOnReviewSource())
	want := loadFixture(t, "blocked-on-review-github")
	assertProviderInputEqual(t, got, want)
}

func TestGitHubAdapterChangesRequested(t *testing.T) {
	got := BuildProviderInput(changesRequestedSource())
	want := loadFixture(t, "changes-requested-github")
	assertProviderInputEqual(t, got, want)
}

func TestGitHubAdapterReviewNotRequired(t *testing.T) {
	got := BuildProviderInput(reviewNotRequiredSource())
	want := loadFixture(t, "review-not-required-github")
	assertProviderInputEqual(t, got, want)
}

func TestGitHubAdapterTrunkIntegrationFailure(t *testing.T) {
	got := BuildProviderInput(trunkIntegrationFailureSource())
	want := loadFixture(t, "trunk-integration-failed-github")
	assertProviderInputEqual(t, got, want)
}

func TestGitHubAdapterPartialEvidence(t *testing.T) {
	got := BuildProviderInput(partialEvidenceSource())
	want := loadFixture(t, "partial-data-github")
	assertProviderInputEqual(t, got, want)
}
